RFCs/0009-Fn::ForEach.md

```diff
@@ -0,0 +1,1461 @@
+    Type: 'AWS::EC2::SubnetNetworkAclAssociation'
+    Properties:
+      SubnetId: !Ref PublicSubnetA
+      NetworkAclId: !Ref NaclPublicSubnet
+  PublicSubnetB:
+    Type: 'AWS::EC2::Subnet'
+    Properties:
+      VpcId: !Ref VPC
+  NaclPublicSubnetBAssociation:
+    Type: 'AWS::EC2::SubnetNetworkAclAssociation'
+    Properties:
+      SubnetId: !Ref PublicSubnetB
+      NetworkAclId: !Ref NaclPublicSubnet
+  PublicSubnetC:
+    Type: 'AWS::EC2::Subnet'
+    Properties:
+      VpcId: !Ref VPC
+  NaclPublicSubnetCAssociation:
+    Type: 'AWS::EC2::SubnetNetworkAclAssociation'
+    Properties:
+      SubnetId: !Ref PublicSubnetC
+      NetworkAclId: !Ref NaclPublicSubnet
+```
+
+### Fn::ForEach in Properties section
+
+#### Use case 4: Replicate similar properties
+
+In this example, customer is using Fn::ForEach to duplicate same inputs to  `AWS::EC2::Instance` properties such as ImageId, InstanceType, and AvailabilityZone.
+
+#### JSON
+
+```json
+{
+    "AWSTemplateFormatVersion": "2010-09-09",
+    "Transform": "AWS::LanguageExtensions",
+    "Mappings": {
+        "InstanceA": {
+            "Properties": {
+                "ImageId": "ami-id1",
+                "InstanceType": "m5.xlarge"
+            }
+        },
+        "InstanceB": {
+            "Properties": {
+                "ImageId": "ami-id2"
+            }
+        },
+        "InstanceC": {
+            "Properties": {
+                "ImageId": "ami-id3",
+                "InstanceType": "m5.2xlarge",
+                "AvailabilityZone": "us-east-1a"
+            }
+        }
+    },
+    "Resources": {
+        "Fn::ForEach::Instances": [
+            "InstanceLogicalId",
+            ["InstanceA", "InstanceB", "InstanceC"],
+            {
+                "${InstanceLogicalId}": {
+                    "Type": "AWS::EC2::Instance",
+                    "Properties": {
+                        "DisableApiTermination": true,
+                        "UserData": {
+                            "Fn::Base64": {
+                                "Fn::Join": ["", [
+                                    "#!/bin/bash\n",
+                                    "yum update -y\n",
+                                    "yum install -y httpd.x86_64\n",
+                                    "systemctl start httpd.service\n",
+                                    "systemctl enable httpd.service\n",
+                                    "echo \"Hello World from $(hostname -f)\" > /var/www/html/index.html\n"
+                                ]]
+                            }
+                        },
+                        "Fn::ForEach::Properties": [
+                            "PropertyName",
+                            ["ImageId", "InstanceType", "AvailabilityZone"],
+                            {
+                                "${PropertyName}": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId"}, "Properties", {"Ref": "PropertyName"}, {"DefaultValue": {"Ref": "AWS::NoValue"}}]}
```

Comment:
Wouldn't it be simpler to use `FindInMap` for each of the attributes, instead of a `ForEach` construct?

```json
"ImageId": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId"}, "Properties", "ImageId", {"DefaultValue": {"Ref": "AWS::NoValue"}},
"InstanceType": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId"}, "Properties", "InstanceType", {"DefaultValue": {"Ref": "AWS::NoValue"}},

```

It's definitely more verbose, but it feels a bit easier to read. The Fn::ForEach here would just be syntax sugar, it wouldn't unblock new use cases.

Reply:
Using `FindInMap` as-is, is a valid alternative, where the power of `Fn::ForEach` in properties would be more evident the more properties there are, as well as the more complex the definition of each property is (see example below)

```json
"ImageId": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId"}, "Properties", "ImageId", {"DefaultValue": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId", "Default", "ImageId"}]}}]}
"InstanceType": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId"}, "Properties", "InstanceType", {"DefaultValue": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId", "Default", "InstanceType"}]}}]}
"AvailabilityZone": {"Fn::FindInMap": [{"Ref": "AvailabilityZone"}, "Properties", "InstanceType", {"DefaultValue": {"Fn::FindInMap": [{"Ref": "InstanceLogicalId", "Default", "AvailabilityZone"}]}}]}
```